lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -307,7 +307,10 @@ impl GraphLinksConverter {
     {
         let edges_len = self.edges.len();
         if level == 0 {
-            (0..edges_len).for_each(|point_id| f(point_id, &self.edges[point_id][0]));
+            (0..edges_len)
+                // Skip points we don't have edges for
+                .filter(|point_id| !self.edges[*point_id].is_empty())
```

Comment:
I'm **not** confident this filter is the correct way to fix the panic that occurred before this change. Without it, an out of bound panic happens for `edges` when building an index with some soft-deleted vectors. I think there may be a deeper problem here. Someone knowledgable may need to look at this with extra care.

I figured that the HNSW graph might have problems when there are gaps between point offsets (happens with soft-deleted vectors). I have not been able to confirm this yet though. The `edges` here originate from links.

The following is very weird to me during investigation: I don't understand the state of `edges` when the panic happens, this can be triggered by removing this filter line above and running the last vacuum test. `edges[0..available_vector_count]` has all non-empty items, while `edges[available_vector_count..total_vector_count]` has some empty items.

Note: when this filter is removed, the last vacuum optimizer test fails with the out of bound panic. That may be a good starting point for any investigation.

Reply:
no, we can't skip points here.